### ISSUE TITLE
IB Gateway on K8s without need for Socat

### ISSUE THREAD

**Issue by tomer-ds** [gnzsnz/ib-gateway-docker#79]
## Is your feature request related to a problem? Please describe

I am running your image in Kubernetes.
As a result there a few patterns that we can use.
Both of them have issues with Socat. 

In one mode Socat sends connections off to ports that are not accesssible due to Kubernetes network management.
In anther mode Socat is unable to function correctly as it can only handle really one connection at a time.

So one way to support this is to run IB-Gateway as a standalone service. In this case we would run Socat as a side-car container separate from your image code. It would be running on the same POD as IB Gateway and therefore all requests come from localhost anyway

The other way is to run IB Gateway as the side car to the requesting service, thereby ensuring that all requests made to IB gateway come from localhost, in which case Socat is no longer needed.

In each case the solution is to disable Socat as long as the image is being run on a Kubernetes POD. 

I could open a PR or fork and make the change for you to merge? Or just leave it up to you?
It seems like such a small change and the idea of me keeping the ability to run this image on Kubernetes all to myself seems selfish

## Describe the solution you'd like

I have already forked your repo and added the required code that would function.
It is a dchoice to run socat based on the existance of a specifically Kubernetes environment var

Here is the new code snippet from the script `latest\scripts\port_forwarding.sh` with some context

```bash
	# set rdp ssh tunnel
	if [ "$GATEWAY_OR_TWS" = "tws" ] && [ -n "$SSH_RDP_PORT" ]; then
		# set ssh tunnel for rdp
		SSH_SCREEN="-R 127.0.0.1:3389:localhost:$SSH_RDP_PORT"
		echo ".> SSH_RDP_TUNNEL set to :${SSH_SCREEN}"
	fi

	if [ -z "$KUBERNETES_SERVICE_HOST" ]; then
		echo ".> Starting socat"
		if [ "$SSH_TUNNEL" = "both" ]; then
			start_socat
		fi
	else
		echo ".> Running in Kubernetes." 
		echo ".> IB will be run as a side car and all requests will come from localhost."
		echo ".> No need for socat."
	fi

	while true; do
		echo ".> Starting ssh tunnel with ssh sock: $SSH_AUTH_SOCK"
		bash -c "ssh ${SSH_ALL_OPTIONS} -TNR 127.0.0.1:${API_PORT}:localhost:${SSH_REMOTE_PORT} ${SSH_SCREEN:-} ${SSH_USER_TUNNEL}"
		sleep "${SSH_RESTART:-5}"
	done
```

I am not saying that this is ideal, probably you can think of a better fit for your process.

In the stable port-forwarding script I did it this way

```bash
else
	if [ -z "$KUBERNETES_SERVICE_HOST" ]; then
		if [ -z "${SOCAT_PORT}" ]; then
			echo ".> SOCAT_PORT not set, port: ${SOCAT_PORT}"
			exit 1
		fi
		if [ -n "$(pgrep -x socat)" ]; then
			# if this script is already running don't start it
			echo ".> socat already active. Not starting a new one"
			exit 0
		fi
		# no ssh tunnel, start socat
		echo ".> Waiting for socat to start"
		sleep 30

		#
		printf "Forking :::%d onto 0.0.0.0:%d > trading mode %s \n" \
			"${API_PORT}" "${SOCAT_PORT}" "${TRADING_MODE}"
		socat TCP-LISTEN:"${SOCAT_PORT}",fork TCP:127.0.0.1:"${API_PORT}"
	else
		echo ".> Running in Kubernetes." 
		echo ".> IB will be run as a side car and all requests will come from localhost."
		echo ".> No need for socat."
	fi
fi
```

## Describe alternatives you've considered

I am sure there is probably a better way

**Comment by gnzsnz** [comment]
hi,

thanks for sharing this. I have a few comments.

First, please review your code base. it's not in sync with master branch. the file that you are proposing to modify is not part of the code base anymore. so this can not be applied as it is.

second, using latest code base you can run the container with SSH_TUNNEL=yes, this will NOT start socat. so if socat is causing you problems, then you can test this option.

lastly, I'm really struggling to understand what is your problem. I would remove socat if i could, it's just one more moving piece that can cause problems, so i'm in favor of removing it. the problem is that it is needed.

TWS api (from ibgateway or TWS desktop) is ONLY exposing the api port on 127.0.0.1, let's say 127.0.0.1:4001 . in order to expose this port to the "outside" it needs to listen on 0.0.0.0:4001, and that can not be done. if you have a way to do it i'm all hears :)

so that's why socat enters in the picture, it will take 127.0.0.1:4001 and exposed it onto 0.0.0.0:4003. once it's there we can publish it to the host. for example the sample docker-compose.yml file is doing "127.0.0.1:4001:4003"

so to recap this last point. you are asking me to disable socat. as i said i'll be happy to do so. please elaborate on how things will work without socat, because there is no way to expose the api(ssh tunnel is one way) to other containers on the docker network nor to the host and beyond. And to be clear NOT being able to access TWS api is not desired.

hope it's clear. looking forward for your comments

**Comment by tomer-ds** [comment]
Very clear.
Many thanks for your insights and comments. I will try with the SSH_TUNNEL option.

When running in Kubernetes we must define the ports exposed by each container at the level of the POD. If a container opens a port or requests that a connection be moved to a new port, but this port is not open in the Kubernetes POD definition then it will not be accessible even if the application opens it on the container.

Socat functionality is based on forking the connections to new ports. When running, each request to port 4003 will be forked to another random port within a range.

```bash
socat TCP-LISTEN:"${SOCAT_PORT}",fork TCP:127.0.0.1:"${API_PORT}"
```
This functionality is based on the ability of the connecting application to access the forked ports, but in Kubernetes this is not possible as they are not truly exposed outside of the container itself.

I tried to remove the `fork` configuration but of course this just meant that once the port was connected by some request, other requests started to fail.

So the 2 patterns that I found that work in Kubernetes are both based on the side car container.

Pattern 1:
- IBGateway is deployed into a POD
- Next to this IBGateway container sits a Socat side car that is not part of the IBGateway image, rather the official Socat image.
- Since both of them are running in the same localhost the requests all come from Socat to IBGateway from 127.0.0.1
- But this pattern is problematic for the same reason as running Socat within the IBGateway container. Forking is not possible

Pattern2:
- The application calling IBGateway is deployed into a POD
- Next to this application container I deploy an IBGateway container as a side car meaning it is in the same POD as the calling application
- All requests coming to IBGateway from the application come from 127.0.0.1 
- This means we do not need Socat at all
- It also means that IBGateway turns into a single-tenant service that can only service the single application running in the same POD
- In this solution I disable the changing of the ports and leave the 400X open via IBGateway config

I am still looking for another way around this, but both of the above solutions require that Socat is disabled within the IBGateway image.

**Comment by tomer-ds** [comment]
I don't know if you have tried this before...
Maybe it is worth looking at?

https://linux.die.net/man/1/redir

**Comment by tomer-ds** [comment]
So it turns out that all this time it was working.
I guess it really was just some config on my side, or maybe something in the latest updates.

Anyway just so you are aware, your image is successfully up and running in our Kubernetes.
